### PR TITLE
Android and iOS. #58 detect changed fingerprints, #186 save/restore credentials

### LINKED ISSUE #58: Detect changed fingerprints
Add check for changes in fingerprint settings. Add event to handle changes on apllication levels and add (possibly optional) ways to recover from changed state.

### For iOS:
Save and compare [evaluatedPolicyDomainState](https://developer.apple.com/documentation/localauthentication/lacontext/1514150-evaluatedpolicydomainstate)

### For Android
Handle `KeyPermanentlyInvalidatedException`. The cipher key becomes invalid of security settings where changed. One way to handle this is given in #54 .

### LINKED ISSUE #186: Save/restore user credentials to/from keychain
### Feature Request
during [#149](https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/issues/149#issuecomment-534160555) discussion

#### Feature Description
That's would be awesome possibility to store encoded user credentials to keychain and restore it.

Same issue here, that would be great to save/retrieve user credential from keychain for few reasons. Firstly first because there's many apps (including mine) where password uses not for auth only, but for encrypting users data (cryptowallets, keychains apps, etc), so in this case your plugin does just a half-deal, another half i should do by myself, encrypting user pass and store it somewhere, this's not so hardcode in there, but would be the awesome feature. Secondly, for security reasons, e.g. changing password when you should confirm old password by urself, or use the fingerprint with keychain to aprrove `Hey, it's me, here my old password` but not only `It's my biometrics here`, another case is banking-app where all sort of user actions should be approved by predefined user pin.

### PR DESCRIPTION
**Pull request by exxbrain**
# Description

1. Fixes #58 for android and iOS.
2. Fixes #186 for android and iOS. 
3. Android biometric updated to 1.0.1.

3 parameters added:

* __secret__: String secret to encrypt and save
* __loadSecret__: If `true` previously saved secret will be loaded and decrypted as a result of successful authentication. 
* __invalidateOnEnrollment__: If `true` secret will be deleted when biometry items are deleted or enrolled 

# How did you test your changes?

Tested manually in android emulator, iPhone SE 13.3 as well as using automated tests. Needs to be tested on a real android device to be sure that #58 completed in android.

### CODE REVIEW

**Comment by NiklasMerz** [comment]
Let me know when I makes sense to review this. Should this be a new major version?

**Comment by exxbrain** [comment]
@NiklasMerz There are no breaking changes, but pretty much functionality added. Probably new major version will be fine.

**Comment by pellaras** [comment]
@exxbrain nice! Thanks for the pull request! Although I would have liked to see new call, it works just fine by passing the extra parameters.

One thing that is bugging me, for iOS, when asking it to store credentials, it does not ask the user to verify first with Touch/Face ID. On android it does. Any thoughts on that?

@NiklasMerz any way we can have this merged?

**Comment by dan0v** [comment]
It would be awesome if this could be merged! The ability to save and load secrets would be greatly appreciated.

File: src/android/BiometricActivity.java [hunk 1]
```diff
@@ -16,15 +16,19 @@
...
 import java.util.concurrent.Executor;
 
+import javax.crypto.Cipher;
+
 public class BiometricActivity extends AppCompatActivity {
 
     private static final int REQUEST_CODE_CONFIRM_DEVICE_CREDENTIALS = 2;
     private PromptInfo mPromptInfo;
+    private CryptographyManager mCryptographyManager;
+    private static final String SECRET_KEY = "__aio_secret_key";
```
Comment (NiklasMerz) [inline src/android/BiometricActivity.java:26]: Should we allow changing this secret key via a parameter?

File: tests/tests.js [hunk 2]
```diff
@@ -75,4 +75,48 @@ exports.defineManualTests = function (contentEl, createActionButton) {
       alert("Authentication invalid " + JSON.stringify(err));
     }
   });
+
+  createActionButton("Save secret", function () {
```
Comment (NiklasMerz) [inline tests/tests.js:79]: We could do these as automatic tests too right? Would be good to test that feature in CI.

File: src/android/Fingerprint.java [hunk 3]
```diff
@@ -20,11 +20,13 @@
...
 
     private static final String TAG = "Fingerprint";
-    private CallbackContext mCallbackContext = null;
-
     private static final int REQUEST_CODE_BIOMETRIC = 1;
+
+    private CallbackContext mCallbackContext = null;
     private PromptInfo.Builder mPromptInfoBuilder;
 
+    static final String SECRET_EXTRA = "secret";
```
Comment (NiklasMerz) [inline src/android/Fingerprint.java:28]: A paramter for this as well?

**Comment by kozmanbalint** [comment]
> It would be awesome if this could be merged! The ability to save and load secrets would be greatly appreciated. In fact, without the ability to save and load secrets, or have some kind of unique token returned (this used to be the case, but `withFingerprint` no longer seems to be returned at all), this plugin cannot be used to securely store user data.

+1 for merging this!
This additional feature is what makes it undoubtly useful in real-world apps.

**Comment by kozmanbalint** [comment]
To move things closer to the acceptance of this PR: an app using this branch has been approved in a fairly strict security audit recently.

**Comment by NiklasMerz** [comment]
Thank you for that Information. That's great to hear. 

I did not have the time to review that properly yet. Looking for any feedback from folks with more native skills than I have.

**Comment by Galiza** [comment]
Would be also much appreciated and super amazing if the Fingerprint.JS for Cordova libs could also include the changes for saving/retrieving the secret.

**Comment by Galiza** [comment]
Also to return the **secret** after **show()** when the user has done the fingerprint authentication. Because if it's a way of us storing this value we want to send it to a backend server for an authentication, and as of now we don't have the value to properly authenticate app side.

File: tests/tests.js (hunk 2 above)
Comment (NiklasMerz) [reply tests/tests.js:79]: We cant't.

**Comment by NiklasMerz** [comment]
I tested this with the new manual tests on my device (Oneplus 6T Android 10) and found two things:

* After saving a secret I need to restart the app. Getting the secret immediately gets me strange encoded string
* __invalidateOnEnrollment__ does not seem to work. I can still get the secret after adding or removing a finger.

I just tested my device with fingerprint only. Possibly other devices are different. If you are reading this please test this branch and report your findings. `cordova plugin add https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio#feature/%2358-detect-changed-fingerprints`

File: src/android/BiometricActivity.java (hunk 1 above)
Comment (Galiza) [reply src/android/BiometricActivity.java:26]: I don't think changing the name of the suffix would be needed, as the SECRET_KEY is only to complement and store more safely on the encryption data.

**Comment by Galiza** [comment]
I was able to find the same issue described by @NiklasMerz and a correct one.

**Case 1.**
Request fingerprint, send _secret_ and _loadSecret_ to the plugin, as return it shows the secret encrypted as described above and as the images below.

**Android Studio**
Screenshot 2020-07-13 at 12 40 35

**Chrome Inspect**
Screenshot 2020-07-13 at 12 46 53

**Case 2.**
After already having the secret saved within the plugin, request the fingerprint prompt, send only _loadSecret_ without sending the secret itself. It returns the secret successfully.

**Comment by NiklasMerz** [comment]
@exxbrain Did you have a chance to look at the reviews? Would love to get this merged sometimes. Otherwise I will have a deeper look.

**Comment by NiklasMerz** [comment]
iOS works as expected

File: src/android/Fingerprint.java (hunk 3 above)
Comment (leolio86400) [reply src/android/Fingerprint.java:28]: It is the key of the map, no point to have a customer value for that.
Not a good practice to have Fingerprint.java using BiometricActivity.java and BiometricActivity.java using Fingerprint.java

Would make sense to move it to PromptInfo.java

**Comment by BorntraegerMarc** [comment]
Is there any ETA for this PR? We are dependent on it for the issue #247

**Comment by leolio86400** [comment]
Seems there is an issue on Android according to the comment above, I'll try to make a prototype using this branch to investigate the issue but might takes time, feel free to help @BorntraegerMarc

**Comment by NiklasMerz** [comment]
I don't give ETAs, sorry. I still need help to fix and test this. Unfortunately I don't have the time and motivation to work on this as much as needed. This plugin really depends on outside contributions and I try to do my best to review, merge and release when everything is in a good shape.

@leolio86400 Thank you for your help. Looking forward. When this issue is resolved I can prepare a release candidate or new version.

**Comment by leolio86400** [comment]
I actually noticed you increased the Cordova android version, that bring a lot of updates to do in my side before even testing 😅

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: src/android/PromptInfo.java [hunk 4]
```diff
@@ -120,18 +122,24 @@ public PromptInfo build() {
...
+            disableBackup = getBooleanArg(argsObject, DISABLE_BACKUP, disableBackup);
+            title = getStringArg(argsObject, TITLE, title);
+            subtitle = getStringArg(argsObject, SUBTITLE, subtitle);
+            description = getStringArg(argsObject, DESCRIPTION, description);
+            fallbackButtonTitle = getStringArg(argsObject, FALLBACK_BUTTON_TITLE, "Use Backup");
+            cancelButtonTitle = getStringArg(argsObject, CANCEL_BUTTON_TITLE, "Cancel");
+            confirmationRequired = getBooleanArg(argsObject, CONFIRMATION_REQUIRED, confirmationRequired);
+            loadSecret = getBooleanArg(LOAD_SECRET, false);
+            invalidateOnEnrollment =getBooleanArg(INVALIDATE_ON_ENROLLMENT, false);
+            secret = getStringArg(SECRET, null);
```
Comment (pellaras) [inline src/android/PromptInfo.java]: @NiklasMerz Should it be the following?

```java
loadSecret = getBooleanArg(argsObject, LOAD_SECRET, false);
invalidateOnEnrollment =getBooleanArg(argsObject, INVALIDATE_ON_ENROLLMENT, false);
secret = getStringArg(argsObject, SECRET, null);
```

**Comment by leolio86400** [comment]
@NiklasMerz so I made a prototype and tested on Android.

I did face some strange thing with `BiometricsActivity` missing from `AndroidManifest.xml` I had to add it by hand (I guess its related to cordova-android upgrade)

Then try 2 scenario on emulator then real test device (Pixel 3XL):

1. Register a user using a simple secret (uuid) and get it

Register:
```show({
      invalidateOnEnrollment: true,
      secret: secret,
    })```

Get secret:
```show({
      invalidateOnEnrollment: true,
      loadSecret: true,
    })```

Worked on both emulator and device, without restarting the app I could get the exact secret I stored.

2. Register a new fingerprint (without restarting the app) and try to get the secret

- Emulator return the secret, not sure if it is expected or not for emulators
- The phone actually return an error as expected 
```{ code: -113
message: "BIOMETRIC_NO_SECRET_FOUND" }```

* I did not have any issue with weird encoding of characters, I guess it is probably due to special characters, would be good to specify that you can only use ascii chars ([a-z0-9\-]+ worked for me), if you have the issue maybe try to do a base64 encoding before and decode when back
* The fingerprint got invalidated as expected after adding a new one

I'll try tomorrow on a low quality device (OPPO) and tell you what is the result but with this result I am only concerned about this weird android manifest issue but that is most likely related to that https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/issues/233 (and seems to be a config.xml issue in my project)

**Comment by leolio86400** [comment]
@NiklasMerz tried on a low end phone and it works as expected too, got the same behaviour as described above 👍

Maybe would make sense to add 2 more methods in the cordova plugin which would be registerBiometricsSecret, and loadBiometricsSecret instead of having everything in one.
It would be more explicit to the developers what you need to do instead of having a behaviour changing depending on parameters and if conditions order. (I can make a PR for that branch but its not a mandatory work I guess)

But in the mean time it looks that it is working as expected on Android, will give it a quick try on iOS.

For the build fail, travis is not detecting the device so I guess it is the CI tool problem not the plugin.

**Comment by leolio86400** [comment]
Tested on iPhone, works as well. So shall we merge that ? Maybe remove the duplicate removeKey in android and all good ?

For the plugin.xml I put a solution [there](https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/issues/233) in case it break for other people too

**Comment by leolio86400** [comment]
@NiklasMerz Hey Nik I created a pull request for the changes I suggested to make more obvious and less error prone https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/pull/268

**Comment by exxbrain** [comment]
@NiklasMerz, It is not compatible with non-androidx anymore. Is it possible to revert compatibility?

**Comment by NiklasMerz** [comment]
@exxbrain What do you suggest doing for the AndroidX thing?

**Comment by exxbrain** [comment]
> @exxbrain What do you suggest doing for the AndroidX thing?

I recovered old branch. It is ok to live it as is for now. Thanks.

**Comment by NiklasMerz** [comment]
@leolio86400 Thanks again for your work. I just did a test on Android with the paramedic tests. I like these two methods for secretes much more.

Does `invalidateOnEnrollment` work properly for you? It does not seem to have an effect for me.

**Comment by leolio86400** [comment]
It does work but not on emulator, what kind of phone do you have ? Android version ? I can search a bit more to find why it is not on your specific device

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by NiklasMerz** [comment]
I tried it on my Oneus 6T. It's running Android 10. The biometric API is implemeted strangely by OP. 
GitHub <notifications@github.com> wrote:
“ It does work but not on emulator, what kind of phone do you have ? Android version ? I can search a bit more to find why it is not on your specific device
 
 Get Outlook for iOS<https://aka.ms/o0ukef>
 ________________________________

**Comment by leolio86400** [comment]
Did some digging, don't find so much information, I can see that we implement it correctly, [Android doc](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder.html#setInvalidatedByBiometricEnrollment(boolean)), similar to these [code samples](https://www.programcreek.com/java-api-examples/?class=android.security.keystore.KeyGenParameterSpec&method=Builder).

To be honest it is not unaware to have random behaviour on Android phones but before going down that road, are you sure your app sample is not doing a registration on opening?

**Comment by NiklasMerz** [comment]
I used `npm run test-android`. Maybe some other test messed it up.

**Review by NiklasMerz — APPROVED** [review]
Just skimming the code it looks good to me. I did not do extensive testing, though.

**Comment by NiklasMerz** [comment]
This has been open for a long time. I am sorry but every time I look at this I get the feeling I just don't know enough of the native platforms to properly review this.

It seems like it is a somewhat stable state right now. Should we merge it and release it and adapt later? Who did test this and would like to see this merged.

**Please add your reaction here** Thumbs up for merging, thumbs down for further reviews.

**Comment by pellaras** [comment]
An early release of this PR is in production for a fairly large organization and it did pass penetration testing. Should be good to go.

**Comment by NiklasMerz** [comment]
I did a test on iOS and it looks like the secret get's loaded without any touchid/faceid prompt. Unfortunately I can only test it on a simulator right now. Does anybody else have this issue?

**Comment by leolio86400** [comment]
> I did a test on iOS and it looks like the secret get's loaded without any touchid/faceid prompt. Unfortunately I can only test it on a simulator right now. Does anybody else have this issue?
> 
> I don't know to much about the Keychain API do debug this properly.

Does it happen that you register and try to load directly ? I think iOS will not prompt biometrics multiple time directly if you do in a short span of time, try to wait something like 5min and try to load again

**Comment by leolio86400** [comment]
Any luck on your testing @NiklasMerz ?

**Review by leolio86400 — APPROVED** [review]

**Comment by sardapv** [comment]
pretty much excited for this feature

**Comment by pataar** [comment]
> I did a test on iOS and it looks like the secret get's loaded without any touchid/faceid prompt. Unfortunately I can only test it on a simulator right now. Does anybody else have this issue?
> 
> I don't know to much about the Keychain API do debug this properly.

Same here on a simulator. On a device it works as expected though.

**Comment by leolio86400** [comment]
@pataar tried on iOS and Android real device and it works, the emulator/simulator does not seems to pick up fingerprint changes.

In Android I do not get prompted if the security settings change and get this error for the first time:
```
{
  code: -113,
  message: "BIOMETRIC_NO_SECRET_FOUND"
}
```

then 
```
{
  code: -100
}
```

**Comment by sardapv** [comment]
any update here?

**Comment by kozmanbalint** [comment]
> any update here?

fingers crossed... :-)

**Comment by NiklasMerz** [comment]
I merged this and will do some basic tests before cutting a release. I think there will probably be bugs in the next release. Please test, report and even better fix them :-)

Thank you all here for your feedback and contributions.

**Comment by kozmanbalint** [comment]
> I merged this and will do some basic tests before cutting a release. I think there will probably be bugs in the next release. Please test, report and even better fix them :-)
> 
> Thank you all here for your feedback and contributions.

Thanks for the merge!!

**Comment by NiklasMerz** [comment]
Tests were a success. Doing the release now!

**Comment by sardapv** [comment]
is it available with an ionic wrapper too or plain cordova?  might be silly question.. sry

**Comment by NiklasMerz** [comment]
It is available for plain cordova and Ionic Native. Ionic Native may need to be updated, I will check that soon.

As always no waranties for this plugin. You may need to check for vulnerabilities yourself.

**Comment by leolio86400** [comment]
We published a fork of the plugin last week and got about 30,000 usage since. It seems to work well, really few errors (or not properly handled exception in the plugin) on Android.

Somehow this one (we do use `isAvailable` before), 17 times (out of 30k)
```
{"code":-100,"message":"java.lang.IllegalStateException: At least one biometric must be enrolled to create keys requiring user authentication for every use"}
```

and a bit more annoying this one, but it happened only once, so I think we'll ignore it for now.
```
{"code":-100,"message":"Keystore operation failed"}
```

One funny that happened once that I have no idea where it comes from since we don't have a error code 3 in the plugin 🤷‍♂️ 
```
{"code":3,"message":"Fingerprint timeout reached. Try again."}
```

Globally, positive results 👍

**Comment by NiklasMerz** [comment]
What do you mean with "we published a fork"? Did you fork this plugin?

**Comment by leolio86400** [comment]
The merge of this was taking too long (hard for a business to justify waiting more than a month on something ready) so we made a copy in our private repo (not publicly published) along with the ionic upgrade to be able to use it. Works nicely 👍 

We will probably use the public version back at some point.

**Comment by NiklasMerz** [comment]
I understand. That makes sense.

**Comment by ifbbprochris** [comment]
@NiklasMerz  when will this feature be released?

**Comment by ifbbprochris** [comment]
@NiklasMerz  
I cant find any code that indicate the biometric change when the user change it in the device level, can you advise?

**Pull request merged**